snowplow-tracker/src/main/java/com/snowplowanalytics/core/tracker/ServiceProvider.kt

```diff
@@ -0,0 +1,342 @@
+            if (configuration is TrackerConfiguration) {
+                trackerConfigurationUpdate.sourceConfig = configuration
+                continue
+            }
+            if (configuration is SubjectConfiguration) {
+                subjectConfigurationUpdate.sourceConfig = configuration
+                continue
+            }
+            if (configuration is SessionConfiguration) {
+                sessionConfigurationUpdate.sourceConfig = configuration
+                continue
+            }
+            if (configuration is EmitterConfiguration) {
+                emitterConfigurationUpdate.sourceConfig = configuration
+                continue
+            }
+            if (configuration is GdprConfiguration) {
+                gdprConfigurationUpdate.sourceConfig = configuration
+                continue
+            }
+            if (configuration is GlobalContextsConfiguration) {
+                globalContextsConfiguration = configuration
+                continue
+            }
+        }
+    }
+
+    private fun stopServices() {
+        tracker?.close()
+        emitter?.shutdown()
+    }
+
+    private fun resetServices() {
+        emitter = null
+        subject = null
+        tracker = null
+    }
+
+    private fun resetControllers() {
+        trackerController = null
+        sessionController = null
+        emitterController = null
+        gdprController = null
+        globalContextsController = null
+        subjectController = null
+        networkController = null
+    }
+
+    private fun resetConfigurationUpdates() {
+        // Don't reset networkConfiguration as it's needed in case it's not passed in the new configurations.
+        // Set a default trackerConfiguration to reset to default if not passed.
+        trackerConfigurationUpdate.sourceConfig = TrackerConfiguration(appId)
+        subjectConfigurationUpdate.sourceConfig = null
+        emitterConfigurationUpdate.sourceConfig = null
+        sessionConfigurationUpdate.sourceConfig = null
+        gdprConfigurationUpdate.sourceConfig = null
+    }
+
+    private fun initializeConfigurationUpdates() {
+        networkConfigurationUpdate = NetworkConfigurationUpdate()
+        trackerConfigurationUpdate = TrackerConfigurationUpdate(appId)
+        emitterConfigurationUpdate = EmitterConfigurationUpdate()
+        subjectConfigurationUpdate = SubjectConfigurationUpdate()
+        sessionConfigurationUpdate = SessionConfigurationUpdate()
+        gdprConfigurationUpdate = GdprConfigurationUpdate()
+    }
+
+    // Getters
+    override fun orMakeSubject(): Subject {
+        return subject ?: makeSubject().also { subject = it }
+    }
+
+    override fun orMakeEmitter(): Emitter {
+        return emitter ?: makeEmitter().also { emitter = it }
+    }
+    
+    override fun orMakeTracker(): Tracker {
+        return tracker ?: makeTracker().also { tracker = it }
+    }
+
+    override fun orMakeTrackerController(): TrackerControllerImpl {
```

Comment:
Could we rename these `orMake...()` functions to `getOrMake...()` as they were before? I guess the `get` part was removed in the translation but the function name doesn't make sense without it.